url2114d537a2_ForNeVeR_lanterna_tgz-pJ8-com_googlecode_lanterna_test_gui_DifferentKindsOfListBoxesTestJ8/src/com/googlecode/lanterna/gui/component/AbstractContainer.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * This file is part of lanterna (http://code.google.com/p/lanterna/).
+ * 
+ * lanterna is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * 
+ * Copyright (C) 2010-2012 Martin
+ */
+
+package com.googlecode.lanterna.gui.component;
+
+import com.googlecode.lanterna.gui.Component;
+import com.googlecode.lanterna.gui.Container;
+import com.googlecode.lanterna.gui.Interactable;
+import com.googlecode.lanterna.gui.InteractableContainer;
+import com.googlecode.lanterna.gui.layout.LayoutParameter;
+import com.googlecode.lanterna.gui.listener.ContainerListener;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ *
+ * @author Martin
+ */
+public abstract class AbstractContainer extends AbstractComponent implements InteractableContainer, Container
+{
+    private final List<ContainerListener> containerListeners;
+    private final List<Component> components;
+
+    protected AbstractContainer()
+    {
+        components = new ArrayList<Component>();
+        containerListeners = new LinkedList<ContainerListener>();
+    }
+
+    @Override
+    public void addComponent(Component component, LayoutParameter... layoutParameters)
+    {
+        if(component == null)
+            return;
+        
+        synchronized(components) {
+            components.add(component);
+        }
+
+        if(component instanceof AbstractComponent)
+            ((AbstractComponent)component).setParent(this);
+    }
+
+    @Override
+    public Component getComponentAt(int index)
+    {
+        synchronized(components) {
+            return components.get(index);
+        }
+    }
+
+    @Override
+    public int getComponentCount()
+    {
+        synchronized(components) {
+           return components.size();
+        }
+    }
+
+    @Override
+    public void removeComponent(Component component)
+    {
+        if(component == null)
+            return;
+        
+        synchronized(components) {
+            components.remove(component);
+        }
+    }
+
+    public void removeAllComponents()
+    {
+        synchronized(components) {
+            while(getComponentCount() > 0)
+                removeComponent(getComponentAt(0));
+        }
+    }
+
+    @Override
+    public boolean isScrollable() {
+        for(Component component: components)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **99**. At this line the following field is not protected by synchronization: `components`. Your code typically protects this field with synchronization on object: `components`.

**Fix**  
 Consider protecting the above field with synchronization on object: `components`. For example, like at lines: **55**, **85**, **74**, and **66**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **55**, **66**, **74**, and **85**.